client/client-multi/client-c/src/kaa/collections/kaa_list.c

```diff
@@ -211,29 +211,15 @@ kaa_error_t kaa_list_remove_first(kaa_list_t *list, match_predicate pred, void *
     KAA_RETURN_IF_NIL3(list, pred, list->size, KAA_ERR_BADPARAM);
 
     kaa_list_node_t *it = kaa_list_find_next(kaa_list_begin(list), pred, context);
-    if (it) {
-        if (it == list->head) {
-            list->head = it->next;
-        }
-        if (it == list->tail) {
-            list->tail = list->tail->prev;
-        }
-
-        set_next_neighbor(it->prev, it->next);
-        destroy_node(it, deallocator);
-        --list->size;
-
-        return KAA_ERR_NONE;
-    }
-
-    return KAA_ERR_NOT_FOUND;
+    kaa_list_remove_at(list, it, deallocator);
+    return it ? KAA_ERR_NONE : KAA_ERR_NOT_FOUND;
 }
 
 void kaa_list_set_data_at(kaa_list_node_t *it, void *data, deallocate_list_data deallocator)
 {
     KAA_RETURN_IF_NIL(it, );
     if (deallocator) {
-        (*deallocator)(it->data);
+        deallocator(it->data);
```

Comment:
I'm not a fan of testing the compiler. (*deallocator)(x) is guaranteed to be the same thing as to deallocator(x).